### PR TITLE
Support host-subnet networking

### PR DESCRIPTION
**Pull request by niusmallnan**
link https://github.com/rancher/rancher/issues/9153

### CODE REVIEW

File: cniconf/watcher.go [hunk 1]
```diff
@@ -11,13 +11,18 @@ import (
...
 	"github.com/rancher/cniglue"
 	"github.com/rancher/go-rancher-metadata/metadata"
 )
 
 var (
-	reapplyEvery = 5 * time.Minute
-	cniDir       = "/etc/cni/%s.d"
+	reapplyEvery        = 5 * time.Minute
+	cniDir              = "/etc/cni/%s.d"
+	hostSubnetLabel     = "io.rancher.host.subnet"
```
Comment (ibuildthecloud) [inline cniconf/watcher.go]: Can you change these to start with io.rancher.network.host

File: cniconf/watcher.go [hunk 2]
```diff
@@ -78,7 +83,39 @@ func (w *watcher) onChange(version string) error {
 	return nil
 }
 
-func (w *watcher) apply(network metadata.Network) error {
+func (w *watcher) inserthostSubnetRules(host metadata.Host, content []byte) ([]byte, error) {
+	subnet, _ := host.Labels[hostSubnetLabel]
```
Comment (ibuildthecloud) [inline cniconf/watcher.go]: These settings should only be applied if the networking type is also host-subnet.  For example, if I have vxlan or ipsec networking turned on, you shouldn't inject these settings if the labels are on the host.

File: trash.conf [hunk 3]
```diff
@@ -18,6 +18,7 @@ github.com/urfave/cli	v1.18.0-84-g3eb41f1
 golang.org/x/net	4cfeeeb
 golang.org/x/sys	c200b10
 github.com/jpillora/backoff a6f6428
+github.com/bitly/go-simplejson da1a892
```
Comment (ibuildthecloud) [inline trash.conf]: When adding new dependencies please use a tag if possible, in this case `v0.5.0` is what you want.

File: cniconf/watcher.go (hunk 1 above)
Comment (niusmallnan) [reply cniconf/watcher.go]: How about `io.rancher.network.host_subnet.xxx`

**Comment by leodotcloud** [comment]
@sheng-liang @ibuildthecloud should we also add support for blacklist range and list of blacklist IPs from the pool? This is useful if the subnet is same as the host subnet and the user wants to blacklist a few IPs.

File: cniconf/watcher.go (hunk 2 above)
Comment (leodotcloud) [reply cniconf/watcher.go]: @ibuildthecloud There are few options here:
1) Use another label to identify `per_host_subnet` network type say `io.rancher.network.per_host_subnet: true` or
2) Add the check based on metadata/network/name == "per-host-subnet" or
3) Add it in the CNI config/name: "rancher-per-host-subnet-cni-network"

Which one to pick?

Comment (leodotcloud) [reply cniconf/watcher.go]: @niusmallnan Let's go with option 1. So that in future if we want to support per_host_subnet on ipsec/vxlan we can support that too.

**Comment by leodotcloud** [comment]
@niusmallnan Two minor things to resolve, otherwise LGTM.

- Addressing Darren's comment, let's choose option 1 as I suggested. If the label is present and if it's true, then we should take rest of the labels into account. This label needs to be added in the catalog of per_host_subnet. Since this label is not present in ipsec/vxlan catalogs these changes don't apply.
- The actual label naming. I prefer using `per_host_subnet` instead of `host_subnet`. Let's wait for the discussion on Slack to finalize on this.

File: cniconf/watcher.go (hunk 2 above)
Comment (niusmallnan) [reply cniconf/watcher.go]: @leodotcloud Where do I set the label? If it is set on the CNI container, it cannot correspond to the managed network. Because the CNI container uses net=host.

Comment (leodotcloud) [reply cniconf/watcher.go]: @niusmallnan I just spoke with @ibuildthecloud and he came up with a totally different alternative.
So for the per-host-subnet catalog template, we will have:
```
subnet: __host_label__: io.rancher.network.per_host_subnet.subnet
rangeStart: __host_label__: io.rancher.network.per_host_subnet.range_start
```

When going through the CNI config, look at the value of the keys before generating the final CNI config, search for `__host_label__`, if found use the label following that to look up the host labels and get the value to populate it.

**Review by leodotcloud — CHANGES_REQUESTED** [review]

File: hostnat/watcher.go [hunk 4]
```diff
@@ -120,6 +126,14 @@ func (w *watcher) networkToRule(network metadata.Network) *MASQRule {
 		bridge, _ := props["bridge"].(string)
```
Comment (leodotcloud) [inline hostnat/watcher.go]: @niusmallnan 

Pseudo code:

```
for aKey, aValue in props {
    if "__host_label__" in aValue {
        label = string.Split(:)
        props[aKey] = hostLabels[label]
    }
}
```

**Review by leodotcloud — CHANGES_REQUESTED** [review]

File: network/per_host_subnet.go [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+	"encoding/json"
+	"strings"
+
+	simplejson "github.com/bitly/go-simplejson"
+	"github.com/pkg/errors"
+	"github.com/rancher/go-rancher-metadata/metadata"
+)
+
+const (
+	PerHostSubnetKey = "__host_label__"
```
Comment (leodotcloud) [inline network/per_host_subnet.go]: This keyword is not related to per host subnet. It can be used with anything else. It's just a keyword to specify we want the value to be replaced with the value of the host label.

File: cniconf/watcher.go [hunk 6]
```diff
@@ -125,3 +127,29 @@ func (w *watcher) apply(network metadata.Network) error {
...
+		return config
+	}
+
+	for aKey, aValue := range props {
+		v, isString := aValue.(string)
+		if !isString {
+			props[aKey] = UpdateCNIConfigByHost(aValue, host)
+		} else {
+			if strings.Contains(v, "__host_label__") {
+				label := strings.TrimSpace(strings.Split(v, ":")[1])
```
Comment (leodotcloud) [inline cniconf/watcher.go]: Please check the length before accessing the array by index. This can potentially cause a crash if there is a problem with config.

File: cniconf/watcher.go [hunk 7]
```diff
@@ -125,3 +127,29 @@ func (w *watcher) apply(network metadata.Network) error {
...
+	if !isMap {
+		return config
+	}
+
+	for aKey, aValue := range props {
+		v, isString := aValue.(string)
+		if !isString {
+			props[aKey] = UpdateCNIConfigByHost(aValue, host)
+		} else {
+			if strings.Contains(v, "__host_label__") {
```
Comment (leodotcloud) [inline cniconf/watcher.go]: `__host_label__`, please make this a const at the top of the file.
`HostLabelKeyword = "__host_label__"`

File: cniconf/watcher.go [hunk 8]
```diff
@@ -125,3 +127,29 @@ func (w *watcher) apply(network metadata.Network) error {
 
 	return lastErr
 }
+
+func UpdateCNIConfigByHost(config interface{}, host metadata.Host) interface{} {
```
Comment (leodotcloud) [inline cniconf/watcher.go]: Giving this function a generic name would be helpful if we introduce more keywords in future.
Say: `UpdateCNIConfigByKeywords`?

**Review by leodotcloud — CHANGES_REQUESTED** [review]

**Comment by leodotcloud** [comment]
Almost LGTM, except for the final few comments mentioned above.

**Comment by niusmallnan** [comment]
@leodotcloud  Have done it. Please recheck.

**Comment by leodotcloud** [comment]
LGTM. @ibuildthecloud can you please take a look?

File: cniconf/watcher.go [hunk 9]
```diff
@@ -125,3 +128,30 @@ func (w *watcher) apply(network metadata.Network) error {
...
+			props[aKey] = UpdateCNIConfigByKeywords(aValue, host)
+		} else {
+			tempSlice := strings.Split(v, ":")
+			if strings.Contains(v, hostLabelKeyword) && len(tempSlice) > 1 {
+				label := strings.TrimSpace(tempSlice[1])
+				lValue := host.Labels[label]
+				if lValue != "" {
+					props[aKey] = lValue
+				} else {
+					delete(props, aKey)
```
Comment (ibuildthecloud) [inline cniconf/watcher.go:150]: I would just replace with empty string

File: cniconf/watcher.go [hunk 10]
```diff
@@ -125,3 +128,30 @@ func (w *watcher) apply(network metadata.Network) error {
...
+	if !isMap {
+		return config
+	}
+
+	for aKey, aValue := range props {
+		v, isString := aValue.(string)
+		if !isString {
+			props[aKey] = UpdateCNIConfigByKeywords(aValue, host)
+		} else {
+			tempSlice := strings.Split(v, ":")
```
Comment (ibuildthecloud) [inline cniconf/watcher.go:143]: this should be SplitN(v, ":", 2) so that labels with `:` in it don't cause issues

File: cniconf/watcher.go [hunk 11]
```diff
@@ -125,3 +128,30 @@ func (w *watcher) apply(network metadata.Network) error {
...
 }
+
+func UpdateCNIConfigByKeywords(config interface{}, host metadata.Host) interface{} {
+	props, isMap := config.(map[string]interface{})
+	if !isMap {
+		return config
+	}
+
+	for aKey, aValue := range props {
+		v, isString := aValue.(string)
```
Comment (ibuildthecloud) [inline cniconf/watcher.go:139]: a more idiomatic approach is

```go
if v, ok := aValue.(string); ok {
  ....
} else {
  ...
}
```

**Comment by cjellick** [comment]
closing in favor of https://github.com/rancher/plugin-manager/pull/84
